pkg/registry/rbac/validation/policy_comparator.go

```diff
@@ -171,3 +176,176 @@ func ruleCovers(ownerRule, subRule rbac.PolicyRule) bool {
+	// appendFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   0  |       1  |       0
+
+	shouldJump := func(rule rbac.PolicyRule) bool {
+		return len(rule.NonResourceURLs) > 0 || len(rule.APIGroups) != 1
+	}
+	c.rules = collapse(c.rules, 5, 2, shouldJump)
+	return c
+}
+func (c *collapser) collapseResrouce() *collapser {
+	// keyFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   1  |       1  |       0
+	// appendFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   0  |       0  |       1
+
+	shouldJump := func(rule rbac.PolicyRule) bool {
+		return len(rule.NonResourceURLs) > 0 || len(rule.Resources) != 1 || len(rule.ResourceNames) != 0
+	}
+	c.rules = collapse(c.rules, 6, 1, shouldJump)
+	return c
+}
+
+func (c *collapser) collapseNonResourceRules() *collapser {
+	// keyFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   1  |       0  |       0
+	// appendFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              1  |   0  |       0  |       0
+
+	shouldJump := func(rule rbac.PolicyRule) bool {
+		return len(rule.NonResourceURLs) != 1
+	}
+	c.rules = collapse(c.rules, 4, 8, shouldJump)
+
+	// keyFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              1  |   0  |       0  |       0
+	// appendFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   1  |       0  |       0
+
+	shouldJump = func(rule rbac.PolicyRule) bool {
+		return len(rule.Verbs) != 1
+	}
+	c.rules = collapse(c.rules, 8, 4, shouldJump)
+
+	return c
+}
+
+func collapse(rules []rbac.PolicyRule, keyFilter, appendFilter int, shouldJump func(rule rbac.PolicyRule) bool) []rbac.PolicyRule {
+	// for example, if keyFilter == 4
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   1  |       0  |       0
+	// It means we choose Verbs as key
+	// AppendFilter decides which field should be collapsed
+
+	compacted := make([]rbac.PolicyRule, 0, len(rules))
+	pool := map[string]*rbac.PolicyRule{}
+	for _, rule := range rules {
+		if shouldJump(rule) {
+			compacted = append(compacted, rule)
+			continue
+		}
+		sortPolicyRule(&rule)
+		key := generatorKey(rule, keyFilter)
+		if existingRule, ok := pool[key]; ok {
+			appendField(existingRule, rule, appendFilter)
+		} else {
+			pool[key] = rule.DeepCopy()
+		}
+	}
+
+	for _, rule := range pool {
+		compacted = append(compacted, *rule)
+	}
+	return compacted
+}
+
+func generatorKey(rule rbac.PolicyRule, filter int) string {
```

Comment:
this assumes no fields will be added to the PolicyRule struct and ignores them when determining a match if new fields are added

Reply:
@liggitt I aggree with you.
If our object is collapsing a group of rules into minimum version, https://github.com/liggitt/audit2rbac/blob/master/pkg/util.go#L44 is not a mature solution. It still needs improvements.
But if what we need here is just a usable collapsed groups, I think your solution is  close to  be mature.
So , please inform me which version do we really need here.
And do you mind if I give a example code which follows your idea.
If not, I will work on this.
Thanks